interp: cannot initialize composite map with a runtime func value

While reviewing #1097 and thinking about additional cases, we found this one:

The following program `sample.go` triggers a panic:

```go
package main

import "strings"

func f(s string) string { return "hello " + s }

func g(s string) string { return "hi " + s }

var methods = map[string]func(string) string{
	"f": f,
	"h": strings.ToLower,
}

func main() {
	methods["i"] = strings.ToUpper
	methods["g"] = g
	println(methods["f"]("test"))
	println(methods["g"]("test"))
	println(methods["h"]("TEST"))
	println(methods["i"]("test"))
}

// Output:
// hello test
// hi test
// test
// TEST
```

Expected result:
```console
$ go run ./sample.go
// hello test
// hi test
// test
// TEST
```

Got:
```console
% yaegi run ./sample.go 
_test/map30.go:9:15: panic
run: reflect.Value.SetMapIndex: value of type func(string) string is not assignable to type *interp.node
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc00018ee00, 0x40)
	/Users/mpl/go1/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc0004a3c68)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:494 +0xc8
panic(0x1a0ac80, 0xc0001b9b10)
	/Users/mpl/go1/src/runtime/panic.go:965 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc00015eb00, 0xc00048b600, 0xc0004a3a48)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:185 +0x253
panic(0x1a0ac80, 0xc0001b9b10)
	/Users/mpl/go1/src/runtime/panic.go:965 +0x1b9
reflect.Value.assignTo(0x1a1e6e0, 0x1b87b70, 0x13, 0x1b5e846, 0x19, 0x1ae5ee0, 0x0, 0x1a0ac80, 0xc0001b95c0, 0x98)
	/Users/mpl/go1/src/reflect/value.go:2451 +0x3f7
reflect.Value.SetMapIndex(0xc00013cf00, 0xc000437f80, 0x15, 0x1a0ac80, 0xc0001b95c0, 0x98, 0x1a1e6e0, 0x1b87b70, 0x13)
	/Users/mpl/go1/src/reflect/value.go:1703 +0x1ec
github.com/traefik/yaegi/interp.mapLit.func1(0xc00015eb00, 0xc00013d140)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:2373 +0x145
github.com/traefik/yaegi/interp.runCfg(0xc00048b600, 0xc00015eb00)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:191 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc000376000, 0xc00048f100, 0x0)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:122 +0x2b0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc000376000, 0xc0003125a0, 0x1c3, 0x7ffeefbff6a2, 0xe, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:583 +0x7d2
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc000376000, 0x7ffeefbff6a2, 0xe, 0xc000312300, 0x1c3, 0x0, 0x0, 0xc0002cc9c0)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:408 +0x105
main.runFile(0xc000376000, 0x7ffeefbff6a2, 0xe, 0x1, 0x1)
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc00011a1a0, 0x1, 0x1, 0xc00010c101, 0xc000102480)
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x86e
main.main()
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:133 +0x479
```

--- a/_test/map30.go
+++ b/_test/map30.go
@@ -6,7 +6,10 @@ func f(s string) string { return "hello " + s }
 
 func g(s string) string { return "hi " + s }
 
-var methods = map[string]func(string) string{"f": f}
+var methods = map[string]func(string) string{
+	"f": f,
+	"h": strings.ToLower,
+}
 
 func main() {
 	methods["i"] = strings.ToUpper
@@ -14,9 +17,11 @@ func main() {
 	println(methods["f"]("test"))
 	println(methods["g"]("test"))
 	println(methods["i"]("test"))
+	println(methods["h"]("TEST"))
 }
 
 // Output:
 // hello test
 // hi test
 // TEST
+// test

--- a/interp/run.go
+++ b/interp/run.go
@@ -2278,7 +2278,6 @@ func arrayLit(n *node) {
 
 	values := make([]func(*frame) reflect.Value, len(child))
 	index := make([]int, len(child))
-	// rtype := n.typ.val.TypeOf()
 	var max, prev int
 
 	for i, c := range child {

--- a/interp/run.go
+++ b/interp/run.go
@@ -570,7 +570,7 @@ func isRecursiveType(t *itype, rtype reflect.Type) bool {
 		return true
 	}
 	switch t.cat {
-	case arrayT, mapT, ptrT, sliceT:
+	case aliasT, arrayT, mapT, ptrT, sliceT:
 		return isRecursiveType(t.val, t.val.rtype)
 	default:
 		return false
@@ -589,34 +589,10 @@ func assign(n *node) {
 
 	for i := 0; i < n.nleft; i++ {
 		dest, src := n.child[i], n.child[sbase+i]
-		switch {
-		case isInterfaceSrc(dest.typ):
-			if len(dest.typ.field) > 0 {
-				svalue[i] = genValueInterface(src)
-				break
-			}
-			svalue[i] = genValue(src)
-		case (dest.typ.cat == valueT || dest.typ.cat == errorT) && dest.typ.rtype.Kind() == reflect.Interface:
-			svalue[i] = genInterfaceWrapper(src, dest.typ.rtype)
-		case isFuncSrc(src.typ) && dest.typ.cat == valueT:
+		if isFuncSrc(src.typ) && isField(dest) {
 			svalue[i] = genFunctionWrapper(src)
-		case isFuncSrc(src.typ) && isField(dest):
-			svalue[i] = genFunctionWrapper(src)
-		case isFuncSrc(dest.typ) && src.typ.cat == valueT:
-			svalue[i] = genValueNode(src)
-		case src.kind == basicLit && src.val == nil:
-			t := dest.typ.TypeOf()
-			svalue[i] = func(*frame) reflect.Value { return reflect.New(t).Elem() }
-		case isRecursiveType(dest.typ, dest.typ.rtype):
-			svalue[i] = genValueRecursiveInterface(src, dest.typ.rtype)
-		case isRecursiveType(src.typ, src.typ.rtype):
-			svalue[i] = genValueRecursiveInterfacePtrValue(src)
-		case src.typ.untyped && isComplex(dest.typ.TypeOf()):
-			svalue[i] = genValueComplex(src)
-		case src.typ.untyped && !dest.typ.untyped:
-			svalue[i] = genValueAs(src, dest.typ.TypeOf())
-		default:
-			svalue[i] = genValue(src)
+		} else {
+			svalue[i] = genDestValue(dest.typ, src)
 		}
 		if isMapEntry(dest) {
 			if isInterfaceSrc(dest.child[1].typ) { // key
@@ -2302,25 +2278,15 @@ func arrayLit(n *node) {
 
 	values := make([]func(*frame) reflect.Value, len(child))
 	index := make([]int, len(child))
-	rtype := n.typ.val.TypeOf()
+	// rtype := n.typ.val.TypeOf()
 	var max, prev int
 
 	for i, c := range child {
 		if c.kind == keyValueExpr {
-			convertLiteralValue(c.child[1], rtype)
-			if isInterfaceSrc(n.typ.val) && !isEmptyInterface(n.typ.val) {
-				values[i] = genValueInterface(c.child[1])
-			} else {
-				values[i] = genValue(c.child[1])
-			}
+			values[i] = genDestValue(n.typ.val, c.child[1])
 			index[i] = int(vInt(c.child[0].rval))
 		} else {
-			convertLiteralValue(c, rtype)
-			if isInterfaceSrc(n.typ.val) && !isEmptyInterface(n.typ.val) {
-				values[i] = genValueInterface(c)
-			} else {
-				values[i] = genValue(c)
-			}
+			values[i] = genDestValue(n.typ.val, c)
 			index[i] = prev
 		}
 		prev = index[i] + 1
@@ -2357,18 +2323,8 @@ func mapLit(n *node) {
 	keys := make([]func(*frame) reflect.Value, len(child))
 	values := make([]func(*frame) reflect.Value, len(child))
 	for i, c := range child {
-		convertLiteralValue(c.child[0], n.typ.key.TypeOf())
-		convertLiteralValue(c.child[1], n.typ.val.TypeOf())
-		if isInterfaceSrc(n.typ.key) {
-			keys[i] = genValueInterface(c.child[0])
-		} else {
-			keys[i] = genValue(c.child[0])
-		}
-		if isInterfaceSrc(n.typ.val) && !isEmptyInterface(n.typ.val) {
-			values[i] = genValueInterface(c.child[1])
-		} else {
-			values[i] = genValue(c.child[1])
-		}
+		keys[i] = genDestValue(n.typ.key, c.child[0])
+		values[i] = genDestValue(n.typ.val, c.child[1])
 	}
 
 	n.exec = func(f *frame) bltn {
@@ -3450,15 +3406,7 @@ func send(n *node) {
 	next := getExec(n.tnext)
 	c0, c1 := n.child[0], n.child[1]
 	value0 := genValue(c0) // Send channel.
-	convertLiteralValue(c1, c0.typ.val.TypeOf())
-
-	var value1 func(*frame) reflect.Value // Value to send.
-	switch {
-	case isInterfaceBin(c0.typ.val):
-		value1 = genInterfaceWrapper(c1, c0.typ.val.rtype)
-	default:
-		value1 = genValue(c1)
-	}
+	value1 := genDestValue(c0.typ.val, c1)
 
 	if !n.interp.cancelChan {
 		// Send is non-cancellable, has the least overhead.

--- a/interp/type.go
+++ b/interp/type.go
@@ -1723,7 +1723,7 @@ func isInterfaceSrc(t *itype) bool {
 }
 
 func isInterfaceBin(t *itype) bool {
-	return t.cat == valueT && t.rtype.Kind() == reflect.Interface
+	return t.cat == valueT && t.rtype.Kind() == reflect.Interface || t.cat == errorT
 }
 
 func isInterface(t *itype) bool {

--- a/interp/value.go
+++ b/interp/value.go
@@ -214,6 +214,31 @@ func genValue(n *node) func(*frame) reflect.Value {
 	}
 }
 
+func genDestValue(typ *itype, n *node) func(*frame) reflect.Value {
+	convertLiteralValue(n, typ.TypeOf())
+	switch {
+	case isInterfaceSrc(typ) && !isEmptyInterface(typ):
+		return genValueInterface(n)
+	case isFuncSrc(typ) && n.typ.cat == valueT:
+		return genValueNode(n)
+	case typ.cat == valueT && isFuncSrc(n.typ):
+		return genFunctionWrapper(n)
+	case isInterfaceBin(typ):
+		return genInterfaceWrapper(n, typ.rtype)
+	case n.kind == basicLit && n.val == nil:
+		return func(*frame) reflect.Value { return reflect.New(typ.rtype).Elem() }
+	case isRecursiveType(typ, typ.rtype):
+		return genValueRecursiveInterface(n, typ.rtype)
+	case isRecursiveType(n.typ, n.typ.rtype):
+		return genValueRecursiveInterfacePtrValue(n)
+	case n.typ.untyped && isComplex(typ.TypeOf()):
+		return genValueComplex(n)
+	case n.typ.untyped && !typ.untyped:
+		return genValueAs(n, typ.TypeOf())
+	}
+	return genValue(n)
+}
+
 func genValueArray(n *node) func(*frame) reflect.Value {
 	value := genValue(n)
 	// dereference array pointer, to support array operations on array pointer
@@ -310,7 +335,7 @@ func zeroInterfaceValue() reflect.Value {
 }
 
 func wantEmptyInterface(n *node) bool {
-	return n.typ.cat == interfaceT && len(n.typ.field) == 0 ||
+	return isEmptyInterface(n.typ) ||
 		n.anc.action == aAssign && n.anc.typ.cat == interfaceT && len(n.anc.typ.field) == 0 ||
 		n.anc.kind == returnStmt && n.anc.val.(*node).typ.ret[0].cat == interfaceT && len(n.anc.val.(*node).typ.ret[0].field) == 0
 }

